### PR TITLE
[Routing] deprecate some router options

### PR DESCRIPTION
**Pull request by Tobion**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | yes 
| Tests pass?   | yes    
| Fixed tickets | https://github.com/symfony/symfony/pull/28865#issuecomment-463480970
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/Routing/Router.php [hunk 1]
```diff
@@ -411,4 +414,15 @@ private function getConfigCacheFactory()
...
     }
+
+    private function checkDeprecatedOption($key)
+    {
+        switch ($key) {
+            case 'generator_base_class':
+            case 'generator_cache_class':
+            case 'matcher_base_class':
+            case 'matcher_cache_class':
+                @trigger_error(sprintf('Option "%s" given to router %s is deprecated since Symfony 4.2.', $key, static::class), E_USER_DEPRECATED);
```
Comment (OskarStark) [inline src/Symfony/Component/Routing/Router.php]: It is deprecated since 4.3 if this goes into `master`, right?

Comment (Tobion) [reply src/Symfony/Component/Routing/Router.php]: good catch

**Comment by Tobion** [comment]
For some reason the non-insulted functional tests are failing.

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -750,17 +752,17 @@ private function registerRouterConfiguration(array $config, ContainerBuilder $co
...
+        $container->setParameter('router.cache_class_prefix', $container->getParameter('kernel.container_class')); // deprecated
         $router = $container->findDefinition('router.default');
         $argument = $router->getArgument(2);
         $argument['strict_requirements'] = $config['strict_requirements'];
         if (isset($config['type'])) {
             $argument['resource_type'] = $config['type'];
         }
         if (!class_exists(CompiledUrlMatcher::class)) {
-            $argument['matcher_class'] = $argument['matcher_base_class'];
+            $argument['matcher_class'] = $argument['matcher_base_class'] = $argument['matcher_base_class'] ?? UrlMatcher::class;
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I need to change this to RedirectableUrlMatcher

**Review by nicolas-grekas — APPROVED** [review]

**Review by stof — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
we should document this in `UPGRADE-4.3.md` and `UPGRADE-5.0.md` too though

File: src/Symfony/Component/Routing/Router.php [hunk 3]
```diff
@@ -411,4 +413,15 @@ private function getConfigCacheFactory()
 
         return $this->configCacheFactory;
     }
+
+    private function checkDeprecatedOption($key)
```
Comment (vudaltsov) [inline src/Symfony/Component/Routing/Router.php]:
```suggestion
    private function checkDeprecatedOption(string $key): void
```

Comment (Tobion) [reply src/Symfony/Component/Routing/Router.php]: that would break the current code for non-string options

Comment (vudaltsov) [reply src/Symfony/Component/Routing/Router.php]: Oh, okay, didn't know they exist. Let's add only `: void` then?

Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Router.php]: to me, `void` is useless, especially on a private method...

**Comment by fabpot** [comment]
Thank you @Tobion.

**Comment by chiqui3d** [comment]
Deprecated? Then how do I know now what cache files Symfony generates for routers?

**Comment by ViPErCZ** [comment]
How prefix UrlGenerator and UrlMatcher in cache directory?

**Comment by ViPErCZ** [comment]
PS: cache_class_prefix is not used now ?

**Comment by flack** [comment]
Hi!

Is there any explanation why these options were being deprecated/removed? I use sf routing standalone in a framework where you have separate application modules (each with its own set of routes) and a prefix that is dynamically changeable (well, at least that's the simplified way to put it). 

The way I implement this is to have one router instance per application module instantiated with the non-prefixed route collection (and before calling `match` I set the prefix with `getContext()->setBaseUrl()`). The ability to specify `matcher_cache_class` and `generator_cache_class` allowed me to cache each of the router instances separately, but with 5.0, I'm now forced to disable caching, which ofc sucks for performance.

Is there any way around this? 

(pls don't say "use Dynamic Router from Symfony CMF". For one thing, it's not compatible with 5.0 yet, also I would have to write a ton of integration code, and by looking at the repo, it seems to be in deep maintenance mode, and I don't want to add a dependency that could get abandoned any time now)

**Comment by flack** [comment]
ok, found a workaround in the linked ticket: You have to set a different `cache_dir` for each router instance now. Would have been nice if this would have been mentioned in UPGRADING.md though

**Comment by xabbuh** [comment]
@flack Would you like to send a PR with the upgrade instruction improvements that you have in mind?

**Comment by flack** [comment]
@xabbuh done in #35297

**Pull request merged**